File: java/21 Merge Two Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        ListNode l3;
        if (l1.val > l2.val) {
            l3 = l2;
            l2 = l2.next;
        }else {
            l3 = l1;
            l1 = l1.next;
        }
        
        ListNode newNode = new ListNode(-1);
        newNode.next = l3;
        while (l1 != null && l2 != null) {
            if (l1.val > l2.val) {
                l3.next = l2;
                l3 = l3.next;
                l2 = l2.next;
            }else {
                l3.next = l1;
                l3 = l3.next;
                l1 = l1.next;
            }
        }
        
        if (l1 != null) {
            l3.next = l1;
        }
        if (l2 != null) {
            l3.next = l2;
        }
        return newNode.next;
    }
}

File: python/21 Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        head = ListNode(0)
        dummy = head
        while l1 and l2:
            if l1.val <= l2.val:
                head.next = l1
                l1 = l1.next
                
            else:
                head.next = l2
                l2 = l2.next
            head = head.next
            
        while l1:
            head.next = l1
            l1 = l1.next
            head = head.next
        while l2:
            head.next = l2
            l2 = l2.next
            head = head.next
        return dummy.next
            
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        if l1 is None:
            return l2
        elif l2 is None:
            return l1
        
        if l1.val <= l2.val:
            l1.next = self.mergeTwoLists(l1.next, l2)
            return l1
        
        else:
            l2.next = self.mergeTwoLists(l1, l2.next)
            return l2
            
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
